text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,608 @@
+>
+> The only browser that actually supports tail calls is Safari (and Webkit). And the Edge team has said that it's unlikely that they will implement tail calls (for similar reasons as Rust: they currently use the Windows ABI calling convention, which doesn't work well with tail calls).
+>
+> Therefore, tail calls in JS is a very controversial thing, even to this day
+>
+> Just to be clear, the Edge team is against implicit tail-calls for all functions, but they're in favor of tail-calls-with-an-explicit-keyword (similar to this RFC).
+
+
+An unofficial summary of the ECMA Script/ Javascript proposal for tail call/return
+https://github.com/carbon-language/carbon-lang/issues/1761#issuecomment-1198672079 (Jul, 2022)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC? -->
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+    - The main uncertainties are regarding the exact restrictions on when backends can guarantee TCO, this RFC is intentionally strict to try and require as little as possible from the backends.
+    - One point that needs to be decided is if TCO should be a feature that needs to be required from all backends or if it can be optional.
+    - Another point that needs to be decided is if TCO is supported by a backend what exactly should be guaranteed? While the guarantee that there is no stack growth should be necessary, should performance (as in transforming `call` instructions into `jmp`) also be guaranteed? Note that a backend that guarantees performance should do so **always** otherwise the main intent of this RFC seems to be lost.
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+    - Are all calling-convention used by Rust available for TCO with the proposed restrictions on function signatures?
+    - Can the restrictions on function signatures be relaxed?
+    - Can generic functions be supported?
+    - Can async functions be supported? (see [here](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-1186604115) for an initial assessment)
+    - Can closures be supported? (see [here](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-1186604115) for an initial assessment)
+    - Can dynamic function calls be supported?
+    - Can functions outside the current crate be supported, functions from dynamically loaded libraries?
+    - Can functions that abort be supported?
+    - Can functions that return a result be supported? As in: `become foo()?;`
```

Comment:
The `?` operator implicitly calls `into()` on the return value, which cannot be supported by a tail call.

Result values are trivially supported - since the caller's return value must match the callee's return value, returning the Result unmodified will simply propagate the Err value or the Ok value.